### ISSUE TITLE
[Plugin] Airflow

### ISSUE THREAD

**Issue by stefanalund** [backstage/backstage#735]
## Summary

Implement the Airflow UI as a Backstage plugin.

## Project website (if applicable)

https://airflow.apache.org/

## Context

Why build a Backstage plugin when there is already a UI for Airflow? The benefit of having a plugin is that Airflow functionality can live together with other infrastructure tools, making it better integrated into the overall developer tooling.

**Comment by JGoldman110** [comment]
@stefanalund I'd like to work on this plugin?

**Comment by stefanalund** [comment]
All yours @JGoldman110 💯 Looking forward to getting more "data" use-cases into Backstage.

**Comment by OrkoHunter** [comment]
Re-opening since plugin suggestion issues are now exempt from [stalebot](https://github.com/backstage/backstage/blob/18f7345a6c29dcfccef7df081b1708f2bf8f62ff/.github/stale.yml#L9).

**Comment by cmpadden** [comment]
Hey @stefanalund and @JGoldman110 , I've started working on an Apache Airflow plugin for Backstage, and while it's minimal right now, I plan to make it more robust when I have cycles here-and-there. At the moment it supports showing metrics on Airflow instance versions, instance health, and a table displaying DAGs and their status. 

I would love if either of you would like to collaborate, contribute, or provide input on the kind of features you would like to see.

https://github.com/cmpadden/backstage-apache-airflow-plugin

<img width="844" alt="Screen Shot 2021-12-02 at 9 21 07 AM" src="https://user-images.githubusercontent.com/5807118/144440123-4862733c-c832-4210-8bd5-e9bf3839e1a3.png">

**Comment by freben** [comment]
@cmpadden Wow that looks great! Seems like it's actually pretty far along - don't hesitate to put up early / not-yet-feature-complete versions and then iterate with more future releases.

**Comment by jrusso1020** [comment]
@cmpadden It might be nice to be able to link to the dag in the airflow UI for now as well for more info? Our data org uses airflow pretty heavily and think it would be impossible to get them completely out of the airflow admin UI but might be nice for everyone if we could bring them into Backstage alongside the rest of our software so they are more visible

**Comment by freben** [comment]
Yeah, replicating visuals / features wholesale (like showing the graphs fully) should usually at most be a stretch goal. Showing summary information and the most critical data points / warnings right where the user needs them in the right entity context, and then linking out to the full experience in an external tool, is often a good balance.

**Comment by cmpadden** [comment]
@jrusso1020 , thanks for the input! I will add some links to the Airflow UI for DAGs. I also plan to add a table of the most-recent DAG runs, and can add relevant links there too.

Sounds good @freben. As for putting up a not-yet-feature-complete version as you mentioned, would you like this to be a pull request into the backstage repository, or should I continue to work in my repo? If the former, I'll do some cleaning up and get that created.

Thanks!

**Comment by freben** [comment]
Pull requests always welcome!

**Comment by cmpadden** [comment]
@jrusso1020 , I've introduced DAG links to the Airflow UI.

If you'd like to follow the pull request, it can be found here: #8348.

**Comment by JGoldman110** [comment]
@cmpadden this looks great! Was waiting on the airflow stable rest API to be rolled out in 2.0. Any thoughts on how the plugin would support running multiple instances of Airflow like DEV, UAT, and PROD or even instances for different business domains?

**Comment by cmpadden** [comment]
> @cmpadden this looks great! Was waiting on the airflow stable rest API to be rolled out in 2.0. Any thoughts on how the plugin would support running multiple instances of Airflow like DEV, UAT, and PROD or even instances for different business domains?

I'll put some thought into supporting multiple Airflow instances, and put it on the road-map of features! It looks like the Jenkins plugin does something similar, in that you can specify a list of instances in the _app-config.yml_ configuration file. It might be a little tricky, however, in how we're using the Backstage proxy plugin to connect to the API; I'm not sure if we'd need to have multiple proxy entries, one for each instance, or if there is a way to have them all route through one.
